src/security-council-mgmt/governors/modules/SecurityCouncilMemberElectionGovernorCountingUpgradeable.sol

```diff
@@ -0,0 +1,230 @@
+
+    /// @dev round => account => weight.
+    ///      weight is the number of tokens/votes cast for the account
+    mapping(uint256 => mapping(address => uint256)) private _weights;
+
+    function __AccountRanker_init(uint256 _maxCandidates) internal onlyInitializing {
+        maxCandidates = _maxCandidates;
+    }
+
+    /// @dev returns the list of top candidates for a given round
+    function _getTopCandidates(uint256 round) internal view returns (address[] memory) {
+        return _candidates[round];
+    }
+
+    /// @dev returns true if the list of top candidates is full for a given round
+    function _isCandidatesListFull(uint256 round) internal view returns (bool) {
+        return _candidates[round].length == maxCandidates;
+    }
+
+    /// @dev returns the weight of an account in a given round
+    function _getWeight(uint256 round, address account) internal view returns (uint256) {
+        return _weights[round][account];
+    }
+
+    /// @dev increases the weight of an account in a given round. 
+    ///      updates the list of top candidates for that round if necessary.
+    function _increaseCandidateWeight(uint256 round, address account, uint256 weightToAdd) internal {
+        address[] storage candidatesPtr = _candidates[round];
+        mapping(address => uint256) storage weightsPtr = _weights[round];
+
+        uint256 oldWeight = weightsPtr[account];
+        uint256 newWeight = oldWeight + weightToAdd;
+
+        // update the weight of account
+        weightsPtr[account] = newWeight;
+
+        // check to see if the account is already in the top K
+        uint256 previousIndexOfAccount = type(uint256).max;
+        // todo: can probably just skip this loop if the oldWeight is less than the weight of the last candidate
+        for (uint256 i = 0; i < candidatesPtr.length; i++) {
+            if (candidatesPtr[i] == account) {
+                previousIndexOfAccount = i;
+                break;
+            }
+        }
+
+        // if the array is not max length yet, and account is not already in the list, just add the account to the end
+        if (previousIndexOfAccount == type(uint256).max && candidatesPtr.length < maxCandidates) {
+            candidatesPtr.push(account);
+            previousIndexOfAccount = candidatesPtr.length - 1;
+        }
+        // if the array is max length already, and account is not already in the list, set the previousIndexOfAccount to the length of the array
+        else if (previousIndexOfAccount == type(uint256).max) {
+            previousIndexOfAccount = candidatesPtr.length;
+        }
+
+        if (previousIndexOfAccount == 0) {
+            // account is already at the top of the list
+            return;
+        }
+
+        // start with the account's index - 1 and move to the left, shifting things down to the right until we find the appropriate spot
+        uint256 j = previousIndexOfAccount - 1;
+        while(true) {
```

Comment:
took my a second, but this looks valid to me; wondering if there's a way to extract this sorting logic into a pure function so it can be easily unit-tested (might not be an organic way to do this, which is fine

Reply:
i think the weight mapping might make it difficult to make pure